src/osd/PrimaryLogPG.cc

```diff
@@ -249,22 +249,30 @@ class PrimaryLogPG::C_OSD_AppliedRecoveredObjectReplica : public Context {
 void PrimaryLogPG::OpContext::start_async_reads(PrimaryLogPG *pg)
 {
   inflightreads = 1;
-  list<pair<boost::tuple<uint64_t, uint64_t, unsigned>,
-	    pair<bufferlist*, Context*> > > in;
+  std::vector<async_read_params_t> in;
   in.swap(pending_async_reads);
   pg->pgbackend->objects_read_async(
     obc->obs.oi.soid,
-    in,
+    std::move(in),
     new OnReadComplete(pg, this), pg->get_pool().fast_read);
 }
 void PrimaryLogPG::OpContext::finish_read(PrimaryLogPG *pg)
 {
+  // TODO(rzarzynski): we're going to need this to be thread-safe.
+  // Currently, BlueStore async_read's completion is always called
+  // from the same bstore_aio thread. However, a thread pool might
+  // become necessary because of decryption/decompression.
   assert(inflightreads > 0);
   --inflightreads;
   if (async_reads_complete()) {
     assert(pg->in_progress_async_reads.size());
-    assert(pg->in_progress_async_reads.front().second == this);
-    pg->in_progress_async_reads.pop_front();
+    // Async reads on replicated pools aren't strictly ordered
+    assert(!pg->pool.info.is_erasure() ||
+           pg->in_progress_async_reads.front().second == this);
+    pg->in_progress_async_reads.remove_if(
+      [this](const std::pair<OpRequestRef, OpContext*>& v) {
+      return v.second == this;
```

Comment:
is this an O(n) search+remove?

Reply:
Yup, the search here is linear. Moreover, as `in_progress_async_reads` is `std::list`, we'll visit all elements which are scattered over memory. Need to find a better way.

P.S.
The branch with sync-vs-async delegation is [here](https://github.com/rzarzynski/ceph/tree/wip-bs-async-delegation). I'm going to push to the PR after cleaning it up.

Reply:
Switched to `std::unordered_map` indexed by `OpContext*`. This should have better complexity at the price of no strict read ordering for EC in `PrimaryLogPG::clear_async_reads`. If that becomes a problem, we can try to restore it there while keeping good characteristic on the main path.